### ISSUE TITLE
cannot put `Constant` type in `U` type --- remove `Constant` wrapper?

### ISSUE THREAD

**Issue by stchang** [stchang/typed-rosette#14]
I expect something like `(Constant Int)` to be a subtype of `(U (Constant Int) (Constant Bool))` but it currently does not hold because `sub?` deconstructs the left type before the right.

But if `sub?` alternatively deconstructs the right before the left, something like `(sub? (Constant Int) Int)` would not hold.

I'm sure we can add special cases to make both cases work, but from my experience so far I think `Constant` causes more problems than it helps so I'm inclined to take it out.

`Constant` is really only used by the solver interface (`forall`, `exists`, `evaluate`, `synthesize`) so I think removing it will have minimal impact.

In contrast, it frequently causes `PosInt`-style "too precise" problems, eg I have many annotations to take away the `Constant` when I initialize a mutable data structure with `Constant` values.

tl;dr: Can I just remove `Constant`?

**Comment by AlexKnauth** [comment]
Yes. Having `forall` etc. accept `Term`s isn't the worst thing.

**Comment by stchang** [comment]
Hmm, won't `Term` have the same problem?

**Comment by stchang** [comment]
For example, I would expect this to work:

```racket
#lang typed/rosette
(let-symbolic [x integer?]
  ((λ ([y : (Term CInt)]) y) (+ x 1)))
```
=>
```racket
; #%app: type mismatch
;   expected:    (Term CInt)
;   given:       Int
;   expressions: (+ x 1)
;   at: (#%app (λ ((y : (Term CInt))) y) (+ x 1))
;   in: (#%app (λ ((y : (Term CInt))) y) (+ x 1))
```

**Comment by AlexKnauth** [comment]
That would if `+` had a case for `(Term CInt)`. The type of `x` is correct, so it works if you do
```racket
(let-symbolic [x integer?]
  ((λ ([y : (Term CInt)]) y) x))
```

Actually, the case
```racket
(C->* [] [] #:rest (Listof Int) (Term CInt))
```
 should be a sound type for `+`. With that the original example would work.

**Comment by stchang** [comment]
Agh, I just noticed that `Int` and friends still use `U`. That makes no sense. I thought `U` means symbolic union? If it doesn't then that is terrible from a usability standpoint.

**Comment by AlexKnauth** [comment]
It uses `U` because it's possible for it to be in a union with something else that happens to be infeasible. `U` means possibly-`union?`.
